libraries/heft-config-file/src/ConfigurationFileLoader.ts

```diff
@@ -0,0 +1,429 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as nodeJsPath from 'path';
+import { JSONPath } from 'jsonpath-plus';
+import { JsonSchema, JsonFile, PackageJsonLookup, Import, FileSystem } from '@rushstack/node-core-library';
+
+interface IConfigurationJson {
+  extends?: string;
+}
+
+/**
+ * @beta
+ */
+export enum InheritanceType {
+  /**
+   * Append additional elements after elements from the parent file's property
+   */
+  append = 'append',
+
+  /**
+   * Discard elements from the parent file's property
+   */
+  replace = 'replace'
+}
+
+/**
+ * @beta
+ */
+export enum PathResolutionMethod {
+  /**
+   * Resolve a path relative to the configuration file
+   */
+  resolvePathRelativeToConfigurationFile,
+
+  /**
+   * Resolve a path relative to the root of the project containing the configuration file
+   */
+  resolvePathRelativeToProjectRoot,
+
+  /**
+   * Treat the property as a NodeJS-style require/import reference and resolve using standard
+   * NodeJS filesystem resolution
+   */
+  NodeResolve
+}
+
+const CONFIGURATION_FILE_FIELD_ANNOTATION: unique symbol = Symbol('configuration-file-field-annotation');
+
+interface IAnnotatedField<TField> {
+  [CONFIGURATION_FILE_FIELD_ANNOTATION]: IConfigurationFileFieldAnnotation<TField>;
+}
+
+interface IConfigurationFileFieldAnnotation<TField> {
+  configurationFilePath: string | undefined;
+  originalValues: { [propertyName in keyof TField]: unknown };
+}
+
+interface IConfigurationFileCacheEntry<TConfigurationFile> {
+  configurationFile?: TConfigurationFile;
+  error?: Error;
+}
+
+/**
+ * @beta
+ */
+export interface IJsonPathMetadata {
+  pathResolutionMethod?: PathResolutionMethod;
+}
+
+/**
+ * @beta
+ */
+export type IPropertyInheritanceTypes<TConfigurationFile> = {
+  [propertyName in keyof TConfigurationFile]?: InheritanceType;
+};
+
+/**
+ * @beta
+ */
+export interface IJsonPathsMetadata {
+  [jsonPath: string]: IJsonPathMetadata;
+}
+
+/**
+ * @beta
+ */
+export interface IConfigurationFileLoaderOptions<TConfigurationFile> {
+  jsonPathMetadata?: IJsonPathsMetadata;
+  propertyInheritanceTypes?: IPropertyInheritanceTypes<TConfigurationFile>;
+}
+
+interface IJsonPathCallbackObject {
+  path: string;
+  parent: object;
+  parentProperty: string;
+  value: string;
+}
+
+/**
+ * @beta
+ */
+export interface IOriginalValueOptions<TParentProperty> {
+  parentObject: TParentProperty;
+  propertyName: keyof TParentProperty;
+}
+
+/**
+ * @beta
+ */
+export class ConfigurationFileLoader<TConfigurationFile> {
```

Comment:
I still suspect that `ConfigurationFile` might be a better name for this class:  If we wanted to modify and resave a config file, wouldn't this class be able to manage that?  Or would we need to introduce a separate class (`PackageJsonEditor`)?